### PR TITLE
Number Spelling

### PR DESCRIPTION
**Pull request by blagasz**
Reinstantiated pull request to make it a clear changeset.

The spelling framework tries to be as generic as possible, but still structured upon my original non-generic implementation of the Hungarian spelling. The Hungarian language in general and the spelling in particular is very complicated, so I hope that the generic helper structure that is suitable for an easy implementation of the Hungarian spelling would be also suitable for all the other languages.

### CODE REVIEW

**Comment by sils1297** [comment]
@blagasz hey, thanks for taking the time to create a PR! Sorry for taking so long to respond. We're trying to get the project up to speed again, could you rebase your commits onto current master? There seem to be conflicts.

**Comment by blagasz** [comment]
Yes, I'll get down to it soon.

Happy to hear from you :)

Szabolcs

**Comment by blagasz** [comment]
Hi @sils1297, I think I managed to rebase first time in my life :)

**Comment by sils1297** [comment]
@blagasz hey, thanks :)

This PR seems to hold some merge commits and other commits related to how you did the rebase due to conflicts I presume. This makes it hard to review it properly and is not really clean with respect to the history of the project.

You seem to be new to git (or at least not that much into the cool stuff it offers ;)) so let me explain how I think this can be solved best :) It's probably best to reset I'm assuming you have the remote `babel` pointing at this repository and `origin` pointing to your fork, if you don't have another remote, just create it with `git remote add babel https://github.com/mitsuhiko/babel/`.

1. Checkout your working branch, which is `master` in your case `git checkout master`
2. Fetch from the babel repository: `git fetch babel`
3. `git diff babel/master` should now show only the changes you want to make.
4. Reset the staging area to the babel master. This will basically reset your git branch to the state of this repositories master **while keeping the state of your files**: `git reset babel/master`
5. A `git diff` should now show the same diff as in 3.
6. You can now use `git add` and `git add -p` to add whole files or partial files to the staging area and `git commit` (without `-a`!) to commit only the staged changes. This way you can create proper logical changeset.
7. You now need to force push. Note that it is not recommended to do such things on master on production repositories! I don't have time to write a full git tutorial here but please keep that in mind. `git push --force origin master`
8. Your clean commits are now on this branch :)

If at any time anything goes wrong - git is your friend and has a history of all recent actions. You can use `git reflog` to access that history and `git reset --hard <commit id>` (attention, destructive action!) to reset the staging area as well as your files to that state.

I hope that helps a bit. If you need anything else, ping me.

**Comment by blagasz** [comment]
Hi @sils1297, thanks for the quick tutorial, I have a long way to go with git :)
Hopefully my second try has what you need. Have a nice day!

**Comment by sils1297** [comment]
@erickwilder can you take a look at this one?
@blagasz cool, that seems to have worked. Could you take a look at http://coala.readthedocs.org/en/latest/Getting_Involved/Writing_Good_Commits/ while we investigate how the code looks. We don't want something with "hopefully works" in the commit message and I think you might find reading this useful.

**Comment by blagasz** [comment]
Sorry about that, it was about the git juggling to clear the commit, not
about the code. I shouldn't have put it in the commit message...

**Comment by sils1297** [comment]
@blagasz yes, please :) `git commit --amend` will let you change the message, no need to go through the reset stuff again in case you don't know. Please try to follow the guidelines on the site I've given, try to summarize the change in the shortlog (first line) and add more elaborate explanation and reasoning in the body.

**Comment by sils1297** [comment]
@blagasz really good commit messages are an art, if you plan to work on open source projects you'll probably learn it sooner or later, don't be frustrated about them or people letting you amend the commit message again and again - you're not the only one struggling with this. Good messages are hard to do and it takes some time to learn to do it properly.

**Comment by blagasz** [comment]
@sils1297 thanks for your support and explanation. I'm not frustrated as I know that everyone is doing this for the greater good! I tried to do my best and any further feedback is very welcome.

**Comment by sils1297** [comment]
@blagasz hey, can you add `Fixes https://github.com/mitsuhiko/babel/issues/179` below your commit message? This way we can track the history of your commits via that bug and can access the old review comments which are lost otherwise.

File: babel/number_spelling.py [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+        for digit in self.digits:
+            yield digit
+
+
+class DigitContext(ContextBase):
+    """
+    Represent one digit of a number and its context
+    """
+    def __init__(self, value, number, side, group, index):
+        ContextBase.__init__(self)
```
Comment (erickwilder) [inline babel/number_spelling.py]: I didn't get why you're not using `super()` here and let Python's inheritance system do it's job.

Comment (sils1297) [inline babel/number_spelling.py]: @erickwilder I personally prefer explicity in those cases, I see only two uses of `super` in Babel so we'll probably want to agree on some style on this. The style guide has no information on this.

File: babel/spelling.py [hunk 2]
```diff
@@ -0,0 +1,549 @@
+# -*- coding: utf-8 -*-
```
Comment (erickwilder) [inline babel/spelling.py]: This entire module redefines a lot of what I've seen at `babel/number_spelling.py`
Can you clean this a little? I suggest deleting some unused code and using a common module for duplicated code (e.g: the exceptions should reside at `babel/exceptions.py`)

File: babel/number_spelling.py (hunk 1 above)
Comment (erickwilder) [inline babel/number_spelling.py]: @sils1297 
I don't see `super()` as an implicit construct of the language. Also, using `super()` makes  easier to do future refactoring that involves changes in the inheritance tree, makes easier to deal with mixins and other very useful patterns.

File: babel/number_spelling.py [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+        number.separator = ' egész '
+
+    if number.rounded:
+        number.prefix = 'kerekítve '
+
+    if number.negative:
+        number.prefix += 'mínusz '
+
+
+    names = (
```
Comment (erickwilder) [inline babel/number_spelling.py]: Sorry. I'm unable to validate this. Any other member of the project speaks Hungarian?

File: tests/test_dates.py [hunk 4]
```diff
@@ -177,13 +177,15 @@ def test_milliseconds_in_day_zero(self):
 
     def test_timezone_rfc822(self):
         tz = timezone('Europe/Berlin')
-        t = tz.localize(datetime(2015, 1, 1, 15, 30))
+        # t = tz.localize(datetime(2015, 1, 1, 15, 30))
```
Comment (erickwilder) [inline tests/test_dates.py]: I think that dead code should remain dead. Remove the comments, please

File: tests/test_number_spelling.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+:copyright: (c) 2013 by Szabolcs Blága
+:license: BSD, see LICENSE in babel for more details.
+"""
+from __future__ import unicode_literals
+
+import pytest
+
+from babel import numbers
+
+def test_public_interface():
```
Comment (erickwilder) [inline tests/test_number_spelling.py]: Can you add the test implementation of this? Also, `test_context_framework` lacks implementation too.

File: tests/test_spelling.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+"""Testing the spelling of numbers
```
Comment (erickwilder) [inline tests/test_spelling.py]: `test_number_spelling.py` and this module seems do be duplicated (or at least very similar). Can you clean up the PR by deleting unused code?

File: babel/number_spelling.py (hunk 1 above)
Comment (sils1297) [inline babel/number_spelling.py]: @erickwilder the thing is `super` provides only one advantage, that is you don't have to do another renaming when you rename the baseclass. However `super` has some caveeats especially when using it for multiple inheritance because you have to make clear which arguments are passed where and such things. That being said, super is complex, I'm sure there are many people out there who don't understand it fully (especially in multiple inheritance cases, I also count me to this group of people btw.) while providing only a very small advantage that is neglectible for many users doing renamings via IDEs. We had some discussions about super usage in general at coala and decided collectively against it. I know that opinions on this are split, your call.

File: babel/spelling.py (hunk 2 above)
Comment (blagasz) [inline babel/spelling.py]: @erickwilder hi, I think I already updated the commit, but maybe it was not succesful. `number_spelling.py` is only a rename of `spelling.py` suggested by @sils1297

File: babel/number_spelling.py (hunk 3 above)
Comment (blagasz) [inline babel/number_spelling.py]: @erickwilder, yes that is locale specific code, there is an ongoing discussion with @sils1297 on how to refactor the code.

File: tests/test_spelling.py (hunk 6 above)
Comment (blagasz) [inline tests/test_spelling.py]: @erickwilder, yes this is also renaming, but I messed something up with the commit --amend or maybe you are not looking at the latest version. When I rename is it necessary, to explicitly `rm` the old file?

Comment (sils1297) [inline tests/test_spelling.py]: @blagasz you can rename with git directly `git mv` or rename on the filesystem, add the new file and commit with the `-a` flag so git recognizes the old file removal.

File: babel/number_spelling.py (hunk 3 above)
Comment (sils1297) [inline babel/number_spelling.py]: I'm unable to validate hungarian. This point is unrelated to the other discussion we had:

I'm of the opinion that if we implement spellers per locale, we'll want to have a directory for that and a file per locale so that locale specific and generic code is clearly seperated and locale specific spelling code can be easily looked up.

@erickwilder Any opinion from your side?

File: babel/number_spelling.py (hunk 1 above)
Comment (erickwilder) [inline babel/number_spelling.py]: What I see is that `super()` itself is not complex. Multiple inheritance on the other hand is. I can't blame `super()` for multiple inheritance complexities (kinda like blame the gun by killing someone, not the killer).

If you're gonna change the way we call parent class methods, this should be done consistently across the library and I think that is is not the subject of this pull request. Just to not extend this discussion, I do think that we can use `super()` here (it's already used, minimally but still used as yourself pointed out) and on a separate thread we can discuss/change the way we handle how to deal with methods overwritten by child classes. 

@sils1297 seems reasonable?

File: babel/number_spelling.py (hunk 3 above)
Comment (erickwilder) [inline babel/number_spelling.py]: I really like the idea of each locale specific code being separated. Since this feature was not in the `master` I would suggest to change this pull request to split the current implementations.
The only thing we need to decide is what conventions to use to achieve such separation without making the code hard to maintain.
@sils1297 @blagasz any suggestions on this?

File: babel/number_spelling.py [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+        """
+        def reverse_group_by(digits, group_size):
+            digits.reverse()
+            groups = []
+            while len(digits) > 0:
+                groups.append(digits[:group_size])
+                digits = digits[group_size:]
+            return groups
+
+        decimal.getcontext().prec = 99 # could be more or less but it should be as many as the supported digits
```
Comment (etanol) [inline babel/number_spelling.py]: This approach is not optimal since it changes the global (per thread) decimal context, which can break user code that relies on decimal context stability, in case a user changes it.

The best way to solve this is to use a decimal context manager. For example:

https://github.com/etanol/babel/commit/7ced0273020427cad3bc321001b64f769cf2f85d

File: tests/test_spelling.py (hunk 6 above)
Comment (erickwilder) [inline tests/test_spelling.py]: @blagasz 
Sorry, I still see two files and I'm looking to the latest revision. Following @sils1297 hints should be enough. 
Looking at you repo, you can see the duplicated files (`number_spelling.py` and `spelling.py`):
screen shot 2015-08-04 at 17 05 37

File: babel/number_spelling.py (hunk 1 above)
Comment (etanol) [inline babel/number_spelling.py]: Just for the record.  `super()` without arguments only works in Python 3.  Python 2 requires two arguments for `super()`.

Highly recommended talk: [Super considered super!  by Raymond Hettinger](https://www.youtube.com/watch?v=EiOglTERPEo)

Comment (blagasz) [inline babel/number_spelling.py]: Ok, I will incorporate all the changes. Should I make a new pull request or
still modify the current commit.

I have nothing against super, I decided to use the simple way to make it
simple, but I could add super no problem.

I also add the factoring of the locale specific code and some loading
mechanism.

I'll leave out decimals global context.

Get rid of unused files.

Get back to you as soon as I have the time to finish all that.

Szabolcs

Comment (etanol) [inline babel/number_spelling.py]: On Tue, Aug 4, 2015 at 10:20 PM, blagasz <notifications@github.com> wrote:
>
> In babel/number_spelling.py:
>
>> +        self.value = sum(d.value*dec(10)**e for e,d in enumerate(self.digits)) # might use int
>> +
>> +    def __len__(self):
>> +        return len(self.digits)
>> +
>> +    def __iter__(self):
>> +        for digit in self.digits:
>> +            yield digit
>> +
>> +
>> +class DigitContext(ContextBase):
>> +    """
>> +    Represent one digit of a number and its context
>> +    """
>> +    def __init__(self, value, number, side, group, index):
>> +        ContextBase.__init__(self)
>
> Ok, I will incorporate all the changes. Should I make a new pull
> request or still modify the current commit.

Since @sils1297 and @erickwilder must be having a hard time sweeping all
the staled pull requests, I believe it is easier for them to just
overwrite the current pull request.

At least, that is what I try to do.  A combination of `git commit
--amend` or `git rebase -i`, followed by `git push -f`.

-- 
Isaac Jurado

"The noblest pleasure is the joy of understanding"
Leonardo da Vinci

Comment (sils1297) [inline babel/number_spelling.py]: @etanol exactly, thanks!

@blagasz please change your commit with `--amend` and the interactive rebase and push force your commits on top. Let me know if you need help.

File: babel/number_spelling.py (hunk 3 above)
Comment (sils1297) [inline babel/number_spelling.py]: I'd make a `number_spelling` directory or so with an `__init__.py` file containing the generic stuff and `de_DE.py` and so on for the locales.

Comment (blagasz) [inline babel/number_spelling.py]: I'm working on it and some questions arise.

Should we factor the testing part into locale specific files as well? If
yes the local specific factoring should be a general rule in the whole
package. Maybe creating locale specific sub-modules and loading them
dynamically could significantly improve speed. Given that in most cases
only a limited set of the locales are used simultaneously.

Coming up with a smooth loading mechanism is not straight-forward so any
idea is welcome. Thanks.

Comment (blagasz) [inline babel/number_spelling.py]: I should sleep one on this as the dynamic loading contradicts locale
negotiation. Other modifications are amended.

File: babel/number_spelling/__init__.py [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+specific function's docstring.
+"""
+from __future__ import division, print_function, unicode_literals
+
+import re, decimal, sys
+
+dec = decimal.Decimal
+
+
+class SpellerNotFound(Exception):
```
Comment (erickwilder) [inline babel/number_spelling/__init__.py]: I still think that exceptions should be in their own module, just like `babel.exceptions` does.

File: babel/number_spelling/__init__.py [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+
+    def __iter__(self):
+        """
+        Not so nice, but this way it could also inherit from ContextBase
+        """
+        yield self.string
+
+
+#####################################
+# locale specific speller functions #
```
Comment (erickwilder) [inline babel/number_spelling/__init__.py]: As far as I can remember @sils1297 had suggested to put each locale specific on isolated modules. Can you split them based on someting like the example below?
- `babel/number_spelling/locale_en.py` - Containing all `en_XX` functions
- `babel/number_spelling/locale_hu.py` - Containing all `hu_XXX` functions

What do you think @sils1297 ?

File: babel/number_spelling/__init__.py [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+        # call transform
+        self._speller(number, ordinal)
+
+        # return the result (construct it recursively)
+        return '{0}'.format(number)
+
+    @classmethod
+    def get_spell_function(cls, locale):
+        """
+        Search `cls.registry` for the spelling function
```
Comment (erickwilder) [inline babel/number_spelling/__init__.py]: Use double back tick to give literal formatting. Single back ticks in RST will make the text appear *emphasised*. E.g:

    ``cls.registry``

File: tests/test_number_spelling.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    assert numbers.spell_number('-0', locale='en_GB') == "minus zero"
+    assert numbers.spell_number(123.25, locale='en_GB') == "one hundred and twenty-three point twenty-five hundredths"
+    assert numbers.spell_number(-12, locale='en_GB') == "minus twelve"
+    assert numbers.spell_number(23457829, locale='en_GB') == "twenty-three million four hundred and fifty-seven thousand eight hundred and twenty-nine"
+    assert numbers.spell_number(1950, locale='en_GB') == "one thousand nine hundred and fifty"
+    assert numbers.spell_number(2001, locale='en_GB') == "two thousand one"
+    assert numbers.spell_number('1999.238', locale='en_GB') == "one thousand nine hundred and ninety-nine point two hundred and thirty-eight thousandths"
+    assert numbers.spell_number(-.199923862, locale='en_GB') == "rounded to minus zero point two tenths"
+    assert numbers.spell_number(-.1, locale='en_GB') == "minus zero point one tenth" # float to string conversion preserves precision
+
```
Comment (erickwilder) [inline tests/test_number_spelling.py]: Insert a new line here. Take a look at the Pocoo Styleguide: http://www.pocoo.org/internal/styleguide/

File: tests/test_number_spelling.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    assert numbers.spell_number(30, True, locale='en_GB') == "thirtieth"
+    assert numbers.spell_number(40, True, locale='en_GB') == "fourtieth"
+    assert numbers.spell_number(123.25, True, locale='en_GB') == "cannot spell fractional numbers as ordinals :("
+    assert numbers.spell_number(-12, True, locale='en_GB') == "minus twelfth"
+    assert numbers.spell_number(23457829, True, locale='en_GB') == "twenty-three million four hundred and fifty-seven thousand eight hundred and twenty-ninth"
+    assert numbers.spell_number(1950, True, locale='en_GB') == "one thousand nine hundred and fiftieth"
+    assert numbers.spell_number(2001, True, locale='en_GB') == "two thousand first"
+    assert numbers.spell_number('1999.23862', True, locale='en_GB') == "cannot spell fractional numbers as ordinals :("
+    assert numbers.spell_number(-.199923862, True, locale='en_GB') == "cannot spell fractional numbers as ordinals :("
+
```
Comment (erickwilder) [inline tests/test_number_spelling.py]: You can remove this unnecessary line

File: tests/test_number_spelling.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    assert numbers.spell_number(3, locale='hu_HU') == "három"
+    assert numbers.spell_number('-0', locale='hu_HU') == "mínusz nulla"
+    assert numbers.spell_number(123.25, locale='hu_HU') == "százhuszonhárom egész huszonöt század"
+    assert numbers.spell_number(-12, locale='hu_HU') == "mínusz tizenkettő"
+    assert numbers.spell_number(23457829, locale='hu_HU') == "huszonhárommillió-négyszázötvenhétezer-nyolcszázhuszonkilenc"
+    assert numbers.spell_number(1950, locale='hu_HU') == "ezerkilencszázötven"
+    assert numbers.spell_number(2001, locale='hu_HU') == "kétezer-egy"
+    assert numbers.spell_number('1999.2386', locale='hu_HU') == "ezerkilencszázkilencvenkilenc egész kétezer-háromszáznyolcvanhat tízezred"
+    assert numbers.spell_number(-.199923862, locale='hu_HU') == "kerekítve mínusz nulla egész százkilencvenkilencezer-kilencszázhuszonnégy milliomod"
+
```
Comment (erickwilder) [inline tests/test_number_spelling.py]: Insert a new line here

File: babel/number_spelling/__init__.py [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+        yield self.string
+
+
+#####################################
+# locale specific speller functions #
+#####################################
+
+def prod(*strings):
+    """generate chartesian products of strings after splitting them"""
+    import itertools
```
Comment (erickwilder) [inline babel/number_spelling/__init__.py]: Can this `import` statement be at the top of the module? `import` inside functions are only acceptable to solve cyclic reference issues and this is not the case

File: babel/number_spelling/__init__.py (hunk 8 above)
Comment (sils1297) [inline babel/number_spelling/__init__.py]: Agreed, this also forces good naming as its not file local.

File: babel/number_spelling/__init__.py (hunk 9 above)
Comment (sils1297) [inline babel/number_spelling/__init__.py]: Sounds good.

Comment (sils1297) [inline babel/number_spelling/__init__.py]: And the tests need to mirror that structure :)

File: babel/number_spelling/__init__.py (hunk 14 above)
Comment (blagasz) [inline babel/number_spelling/__init__.py]: I did this this way because I wanted to show that the implementation of the function is not important and might change in the future (and then the header imports could stay the same). I could move the import if you say the general rule (import at the beginning) is stronger in this case.

File: babel/number_spelling/__init__.py (hunk 9 above)
Comment (blagasz) [inline babel/number_spelling/__init__.py]: @erickwilder and @sils1297 as I mentioned earlier this structure could work only if I import all the submodules when the ``__init__.py`` runs, otherwise locale negotiation would not work. It this case the import mechanism gives a little overhead to the module, but the structure improves. I was thinking about a more elegant solution, but couldn't come up with one yet.

(We should make some connection with the CLDR here, but the number spelling in CLDR is seriously broken. It could be used for flagging which locale has a spelling implementation or we should come up with an internal flagging. Both ways the import mechanism could become more dynamic and thus efficient.)

**Comment by blagasz** [comment]
@sils1297 and @erickwilder, I think ``number_spelling`` should become ``babel.numbers._spelling`` as the only public interface of the entire module is the function ``spell_number`` in the ``numbers`` module.

**Comment by erickwilder** [comment]
@blagasz 
I agree with your last comment. Sorry for this huge delay to interact here.
Despite of the comment, I'm promoting the status of this to `partial approved` - If you don't want to make those changes let me know and we can move forward.

**Comment by erickwilder** [comment]
@blagasz 
One more thing (I promise :smile:) - Can you add a short entry in the `CHANGES` file with a minimal description of the work you've done? This way we can keep track of what's the contents of out next release more easily.

Thanks again.

**Comment by blagasz** [comment]
OK, I make the changes as I really think it is better that way, and also add to ``CHANGES``.

**Comment by blagasz** [comment]
@erickwilder and @sils1297 I need some help as I have no idea what conflicts GitHub is telling me about and how to resolve them. Maybe I need a new rebase?

**Comment by sils1297** [comment]
Yes please rebase. If you are prompted to merge install `meld` and use the
`git mergetool` command to invoke it, should be the easiest way.

**Comment by blagasz** [comment]
@sils1297 rebase ready. If you merge this I will create a new pull request for the documentation, is it OK that way?

**Comment by blagasz** [comment]
@sils1297 and @erickwilder how is it going? If I could help in any other way, let me know!

**Comment by sils1297** [comment]
@blagasz @erickwilder will probably review this soon, he was unavailable before due to personal reasons. We'll let you know if something's missing - sorry for the delay!

**Comment by sils1297** [comment]
Oh @blagasz could you rebase? We merged quite some stuff and it seems there are conflicts now :/

**Comment by blagasz** [comment]
@sils1297, I completed the rebase, but AppVeyor fails, any idea why?

No problem with the delay, I've just wanted to no if I could help some other ways as well...

**Comment by sils1297** [comment]
@blagasz GitHub shows that the branch still has conflicts, that's why CI doesn't run and AppVeyor shows the "failed" build status. Please try: `git fetch mitsuhiko && git rebase mitsuhiko/master` while replacing `mitsuhiko` with whatever alias you use for this original repository. Then push force it up again.

**Comment by blagasz** [comment]
@sils1297 already tried that and git says ``Current branch origin/master is up to date.``. So I really have no clue.

**Comment by blagasz** [comment]
Sorry @sils1297, I messed something up very much, but now I restarted and it works. (I think I rebased the wrong way and then the intended way again or something like that...)

**Comment by sils1297** [comment]
Marking this needs-rebase, github shows conflicts.

**Comment by blagasz** [comment]
:)

I am acquiring great routine in rebasing. It would be easier to rebase
before you actually going to merge this pull as the other pulls you merge
will always necessitate rebase. Anyway, I could do the rebase this evening!

**Comment by sils1297** [comment]
@blagasz it would be easier but if there are conflicts than we'll want to re-review this anyways. I know this sucks a bit but it's needed to make sure we don't introduce bugs through conflict resolving.

**Comment by sils1297** [comment]
@blagasz also I'm sorry that this takes so long, it would of course be less cumbersome if we had the manpower to review this PR faster. We seem to get hold of things slowly after taking over this repo so in the future this will hopefully be better. In case you're interested to help out, please look at https://github.com/python-babel/babel/issues/227 :)

**Comment by blagasz** [comment]
@sils1297 I agree, what I meant is that it is more resource optimal if you do one rebase once before merging each pull. Anyway, I look at #227 and try to help! Also doing the rebase ASAP.

**Comment by blagasz** [comment]
@sils1297, rebase complete. The conflict was only due to the ``CHANGES`` file.

**Comment by blagasz** [comment]
@sils1297 I managed to come up with solutions. The ``ContextMaker`` class is gone and a new more generic ``Preprocessor`` structure is introduced. I was also trying to simplify the framework for locale specific spelling implementations and increase flexibility by allowing custom parameters. I also added some more test cases, especially for the new errors. All the errors are handled in the preprocessor, so locale specific implementation need not bother with that (or they could through custom preprocessors). I wrote some docs as well, but we agreed to make it a separate PR.

Cheers

**Comment by sils1297** [comment]
> but we agreed to make it a separate PR.

Uhm I don't recall, would you mind simply adding the docs here? Just as another commit but reviewing them together makes sense IMO. (At the danger of saying inconsistent things here...)

**Comment by blagasz** [comment]
It is already there in the ``test_framework.py``. It throws a SpellerNotFound error. I could change the name...

**Comment by sils1297** [comment]
> It is already there in the test_framework.py. It throws a SpellerNotFound error. I could change the name...

I don't see any text for that

**Comment by blagasz** [comment]
Sorry I haven't staged that...

**Comment by blagasz** [comment]
@sils1297 I pushed everything now, sorry for the fuss!

**Comment by blagasz** [comment]
Rebased and line removed :)

**Comment by sils1297** [comment]
Cool, thanks :) This is growing very good, many revisions -> high quality :)

**Comment by sils1297** [comment]
Oh @etanol could you please take at least a short look at this? You might have valuable additions.

**Comment by etanol** [comment]
@sils1297  I had this PR merged in my custom branch.  However, it seems to have changed slightly.  I will take a look to the new `locale_*` submodules as I haven't had the time to figure them out.  If there is a more urgent need, go ahead and merge it to *master*; then I can take a look in two weeks to see if something needs to be amended.

**Comment by sils1297** [comment]
@blagasz hey, so I begin to like this code more and more. I think it would be good to have a few more tests for some corner cases. The patch coverage (https://codecov.io/github/python-babel/babel/compare/46a234e...7fbf2c5) shows that there are some cases like some exceptions and a break statement and so on missing, I'd love to see a few more tests for those to make this even more future proof. Those exceptions e.g. should be rather easy to test.

**Comment by sils1297** [comment]
And well there's the conflict again probably because of the CHANGES file so this'll probably reoccur. Maybe wait with the rebase a bit if it's cumbersome for you.

**Comment by sils1297** [comment]
Leaving marked as scheduled for review because I (or somebody else) still needs another deep look at this code and I can't do that *right* now.

**Comment by blagasz** [comment]
@sils1297 I was trying to improve the coverage and hope you'll will still like the code after a deeper look. I'm working on the Chinese simplified spelling right now, but I will put it in a separate pull. Also I am planning to write a short tutorial for the framework to help collaboration to produce implementations for other locals quickly.

**Comment by sils1297** [comment]
@blagasz tests fail. We'll wait with review until this is repaired as this is not eligible for merge as long as any CI fails.

**Comment by blagasz** [comment]
Yes, sorry, I know about the problem, some of the new tests fail on 3.* and
I only tested locally on 2.7. A few minutes and I patch that!

**Comment by blagasz** [comment]
@sils1297 all set :)

**Comment by mitsuhiko** [comment]
I'm a strong -1 on pulling this in.  The reason for this is that Babel ultimately only allows two sources of information in the library: It either has to come from the olson database or CLDR.  Historically there have been only very few exceptions to this rule and they were made for very specific reasons.

The number spelling is specified in the `rbnf` rules (see `common/rbnf/LANG.xml`) in the CLDR dataset and can be implemented through a rule engine.  I would welcome an implementation of this, but definitely not a manually maintained for a few languages in Babel.  It's a maintenance nightmare and impossible to judge.  The only reason Babel works is because people that actually understand what they are doing are maintaining the CLDR dataset :)

**Comment by sils1297** [comment]
@blagasz I do tend to agree with @mitsuhiko . It is ultimately my fault to not have this stopped earlier.

Still the experience you gained might be very helpful understanding and implementing spelling support using CLDR. Are you still interested in working on https://github.com/python-babel/babel/issues/179 ?

**Comment by blagasz** [comment]
@sils1297 and @mitsuhiko no hard feelings at all, but the strong -1 from Armin was a tough one :)

I completely agree with Armin and I was myself a little bit surprised when he said two years ago, that my approach "Sounds like something that can find its way into Babel". I was wondering about the CLDR support and every other possibilities then and I came up with my solution as the only viable alternative. The CLDR rbnf rules sounded interesting but it was insufficient and unimplemented in CLDR two years ago, I checked out the new version and it seems to have improved a lot, now it seems something implementable. It is also my fault that I did not checked that earlier as a long time passed since I started my implementation. Maybe I was in love with my own code :)

Anyway, I started to read the docs for the rule based formatting and I am happy to implement the engine (or transform an existing implementation into Python). I am busier these days, but try to have some advancement in that soon.

Personally I gained a lot by this experience (failure) and I am happy to help in any way in the future.

**Comment by sils1297** [comment]
@blagasz you can accept the invitation for our org on https://github.com/python-babel so I can assign you to the bug. Many thanks for sticking with us.
